Make package compatible with React 16

james-ecargo opened the pull request:
This PR updates the package to make it compatible with either React 15 or 16 by updating references to `React.createClass` and `React.propTypes`.

james-ecargo committed:
Bump React versions for dependencies and peer dependencies

Use the same updated versions of 'react' and 'react-dom' for both
dependencies and peer dependencies.

This removes any confusion over which version is specifically being
targeted and accounts for the change to the versioning scheme in React 15
which increased the major version number from 0 to 15, hence the '15.0.0'
version number.

james-ecargo committed:
Make react-reorderable compatible with React 16

In React 16 the function 'React.createClass' and the 'React.propTypes'
object have been deprecated and moved to the 'create-react-class' and
'prop-types' packages respectively.

In this commit I have updated references to these to point to the new
packages and updated the references in the 'package.json' file.

james-ecargo committed:
Remove ambiguous minified file

The current Gulp build task only creates 'react-reorderable.min.js' which
means that this file is no longer updated when a new version is built.

james-ecargo committed:
Loosen package version restrictions and make into peer dependencies

In order to keep the package as small and flexible as possible I've
followed the current approach of externalizing all the dependencies.

In this commit I've added the 'create-react-class' and 'prop-types'
references to the peer dependencies of the package.

I've also loosened the version number restrictions to make this as widely
compatible as possible.

james-ecargo committed:
Make build task more discoverable

Add the Gulp build entry point as an NPM script. This makes it easier to
discover.

james-ecargo committed:
Externalize peer dependencies

Add the 'create-react-class' and 'prop-types' peer dependencies as
external dependencies in Browserify. This results in a smaller bundle size
and these packages are not included in the final output.

james-ecargo committed:
Make React 16 a possible peer dependency

This package should now be compatible with React 16 so this step allows it
to be included as a peer dependency.

james-ecargo committed:
Specify the correct dependencies

I have already updated the peer dependencies so it pays to have the same
match these to the declared dependencies as well.

mgechev commented:
@james-ecargo awesome job! Would you also add a fix for the demo?

james-ecargo commented:
@mgechev sure, I'll have a look at fixing it.

mgechev commented:
Thank you!

mgechev commented on package.json:
```diff
@@ -26,9 +27,11 @@
+    "prop-types": "^15.0.0",
+    "react": "^15.0.0 || ^16.0.0",
+    "react-dom": "^15.0.0 || ^16.0.0",
+    "react-drag": "^0.3.0"
```
We can now update it to `react-drag@2.0.0`.

mgechev commented:
@james-ecargo added one comment. If we have the demo working I think it's ready to merge.

james-ecargo committed:
Depend on compatible version of 'react-drag' package

'react-drag' >= 2.0.0 now supports React 16 which is what we want to
target in this module.

james-ecargo committed:
Move dependencies to peer dependencies

As we externalize these modules they are only relevant as peer
dependencies.

james-ecargo committed:
Add Browserify shim aliases for new packages

As these packages are expected to the be external we need to add aliases
so they can be loaded in the browser.

james-ecargo committed:
Fix demo

Update versions of React packages and add new packages
'create-react-class' and 'prop-types'.

james-ecargo commented:
@mgechev I have fixed the demo using the production versions of React which appears to work correctly and has no warnings or errors in the browser console.

However if you use the development versions provided with the React packages:
```
diff --git a/demo/index.html b/demo/index.html
index b28edc6..619e4a6 100644
--- a/demo/index.html
+++ b/demo/index.html
@@ -48,8 +48,8 @@
 <div id="content"></div>
 <div id="events">
 </div>
-<script src="../node_modules/react/umd/react.production.min.js"></script>
-<script src="../node_modules/react-dom/umd/react-dom.production.min.js"></script>
+<script src="../node_modules/react/umd/react.development.js"></script>
+<script src="../node_modules/react-dom/umd/react-dom.development.js"></script>
 <script src="../node_modules/prop-types/prop-types.js"></script>
 <script src="../node_modules/create-react-class/create-react-class.js"></script>
 <script src="../node_modules/react-drag/dist/react-drag.js"></script>
```

Then you can observe the following error about assigning to the array of children:
```
Uncaught TypeError: Cannot assign to read only property '0' of object '[object Array]'
    at indexChildren (react-reorderable.js:112)
    at Object.getInitialState (react-reorderable.js:167)
    at new <anonymous> (create-react-class.js:870)
    at constructClassInstance (react-dom.development.js:6897)
    at updateClassComponent (react-dom.development.js:8381)
    at beginWork (react-dom.development.js:8767)
    at performUnitOfWork (react-dom.development.js:10766)
    at workLoop (react-dom.development.js:10830)
    at HTMLUnknownElement.callCallback (react-dom.development.js:580)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:619)
```
...and this one which appears to be fairly straightforward (explained [here](https://reactjs.org/docs/cross-origin-errors.html)):
```
Uncaught Error: A cross-origin error was thrown. React doesn't have access to the actual error object in development. See https://fb.me/react-crossorigin-error for more information.
    at Object.invokeGuardedCallbackDev (react-dom.development.js:626)
    at invokeGuardedCallback (react-dom.development.js:476)
    at renderRoot (react-dom.development.js:10908)
    at performWorkOnRoot (react-dom.development.js:11556)
    at performWork (react-dom.development.js:11509)
    at requestWork (react-dom.development.js:11420)
    at scheduleWorkImpl (react-dom.development.js:11274)
    at scheduleWork (react-dom.development.js:11231)
    at scheduleTopLevelUpdate (react-dom.development.js:11735)
    at Object.updateContainer (react-dom.development.js:11773)
```

I am none the wiser as to why the first error happens in the development versions but doesn't appear to affect the production version.

mgechev commented:
@james-ecargo merged and published under react-reorderable@0.4.0.
